core/src/main/java/org/elasticsearch/cluster/metadata/Snapshot.java

```diff
@@ -78,18 +82,29 @@ public boolean equals(Object o) {
             return false;
         }
         @SuppressWarnings("unchecked") Snapshot that = (Snapshot) o;
-        return repository.equals(that.repository) && name.equals(that.name);
+        return repository.equals(that.repository) && snapshotId.equals(that.snapshotId);
     }
 
     @Override
     public int hashCode() {
-        return Objects.hash(repository, name);
+        return hashCode;
+    }
+
+    private int computeHashCode() {
+        return Objects.hash(repository, snapshotId);
     }
 
     @Override
     public void writeTo(final StreamOutput out) throws IOException {
         out.writeString(repository);
-        out.writeString(name);
+        snapshotId.writeTo(out);
+    }
+
+    /**
+     * Returns whether the snapshot logically equivalent to the given repository name + snapshot name combination.
+     */
+    public boolean isSame(final String repositoryName, final String snapshotName) {
```

Comment:
this is effectively used only in 3 places (due to weird listener completion logic). maybe we can just do equals there and get rid of this method.